subsys/bluetooth/services/cts_client.c

```diff
@@ -0,0 +1,222 @@
+ * |     2 bytes     |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  |1 byte  | = 10 bytes.
+ */
+#define CTS_C_CURRENT_TIME_EXPECTED_LENGTH 10
+
+int bt_cts_client_init(struct bt_cts_client *cts_c,
+		       bt_cts_client_evt_handler evt_handler)
+{
+	int err;
+
+	if (!cts_c || !evt_handler) {
+		return -EINVAL;
+	}
+
+	memset(cts_c, 0, sizeof(struct bt_cts_client));
+
+	cts_c->evt_handler = evt_handler;
+
+	err = k_sem_init(&cts_c->read_params_sem, 1, 1);
+
+	return err;
+}
+
+void bt_cts_client_on_disconnected(struct bt_cts_client *cts_c)
+{
+	k_sem_give(&cts_c->read_params_sem);
+}
+
+int bt_cts_client_handles_assign(struct bt_gatt_dm *dm,
+				 struct bt_cts_client *cts_c)
+{
+	const struct bt_gatt_dm_attr *gatt_service_attr =
+		bt_gatt_dm_service_get(dm);
+	const struct bt_gatt_service_val *gatt_service =
+		bt_gatt_dm_attr_service_val(gatt_service_attr);
+	const struct bt_gatt_dm_attr *gatt_chrc;
+	const struct bt_gatt_dm_attr *gatt_desc;
+
+	if (bt_uuid_cmp(gatt_service->uuid, BT_UUID_CTS)) {
+		return -ENOTSUP;
+	}
+	LOG_DBG("CTS found");
+
+	gatt_chrc = bt_gatt_dm_char_by_uuid(dm, BT_UUID_CTS_CURRENT_TIME);
+	if (!gatt_chrc) {
+		return -EINVAL;
+	}
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc,
+					    BT_UUID_CTS_CURRENT_TIME);
+	if (!gatt_desc) {
+		return -EINVAL;
+	}
+	cts_c->handle_ct = gatt_desc->handle;
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, gatt_chrc, BT_UUID_GATT_CCC);
+	if (!gatt_desc) {
+		return -EINVAL;
+	}
+	cts_c->handle_ct_ccc = gatt_desc->handle;
+	LOG_DBG("Current Time characteristic found.");
+
+	/* Finally - save connection object */
+	cts_c->conn = bt_gatt_dm_conn_get(dm);
+
+	return 0;
+}
+
+/**@brief Function for decoding a read from the Current Time characteristic.
+ *
+ * @param[in] time    Current Time structure.
+ * @param[in] data    Pointer to the buffer containing the Current Time.
+ * @param[in] length  Length of the buffer containing the Current Time.
+ *
+ * @retval 0        If the time struct is valid.
+ * @retval -EINVAL 	If the length does not match the expected size of the data.
+ */
+static int current_time_decode(struct bt_cts_current_time_char *time,
+			       uint8_t const *data, uint32_t const length)
+{
+	if (length != CTS_C_CURRENT_TIME_EXPECTED_LENGTH) {
+		/* Return to prevent accessing the out-of-bounds data. */
+		return -EINVAL;
+	}
+
+	/* Date. */
+	time->exact_time_256.year = sys_get_le16(data);
+	data += sizeof(uint16_t);
+	time->exact_time_256.month = *data++;
+	time->exact_time_256.day = *data++;
+	time->exact_time_256.hours = *data++;
+	time->exact_time_256.minutes = *data++;
+	time->exact_time_256.seconds = *data++;
+
+	/* Day of week. */
+	time->exact_time_256.day_of_week = *data++;
+
+	/* Fractions of a second. */
+	time->exact_time_256.fractions256 = *data++;
+
+	/* Reason for updating the time. */
+	time->adjust_reason.manual_time_update = (*data >> 0) & 0x01;
```

Comment:
Maybe in this line:
`(*data) & BIT(0)`
and next 
`(*data) & BIT(1)`
?

Reply:
Done.
"(*data) & BIT(0)" works.
"(*data) & BIT(1)" does not work as expected. It is modified to "(*data & BIT(1)) ? 1 : 0".

Reply:
Sorry about that